[1단계 - 음식점 목록] - 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
## 🌼 니야와 데이지의 페어 프로그래밍 🌼

안녕하세요 마르코, 니야입니다 :) 이번 `점심 뭐 먹지` 미션 잘 부탁드려요! 😁

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: [🍀 니야의 점심 뭐먹지 웹사이트 보러가기](https://sooyeoniya.github.io/javascript-lunch/)

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

컴포넌트 구조를 설계하는 과정이 가장 어려웠습니다. 특히, 어디까지 추상화를 진행해야 할지가 가장 고민이었습니다.

처음에는 `Button` 컴포넌트를 만들었지만, 이후 두 개의 버튼을 함께 사용하는 `UI`가 존재한다는 점을 고려해 `ButtonGroup` 컴포넌트로 추상화하려고 했습니다. 하지만 이 과정에서 몇 가지 문제가 발생했습니다.

1. `Button` 자체가 `type`, `text`, `action` 세 가지 매개변수를 필요로 했습니다.
2. `ButtonGroup`을 생성할 경우, 두 개의 버튼 각각에 대한 인자를 외부에서 전달해야 하므로, 매개변수 개수가 두 배로 증가했습니다.
3. 게다가, 2단계에서도 사용할 수 있도록 `ButtonGroup`을 만들려다 보니, 위 인자들뿐만 아니라 `onClick`까지 외부에서 주입해야 하는 상황이 발생했습니다.
4. 따라서 총 4개 이상의 인자가 필요해졌고, `RestaurantForm` → `ButtonGroup` → `Button`으로 불필요한 데이터 전달이 발생한다고 판단했습니다.
이러한 이유로, 결국 `ButtonGroup` 컴포넌트 생성을 하지 않기로 결정했습니다. 하지만 2단계 작업을 진행하게 되면 다시 컴포넌트 분리를 고민해봐야 할 것 같다는 생각이 듭니다.

일단 `YAGNI` 원칙에 따라 당장 필요하지 않아서 구현하지 않기로 했지만, 이후 유지보수성과 재사용성을 고려해 다시 한 번 고민해봐야 할 것 같습니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

**✅ 각 컴포넌트의 렌더링 방식 개선 고민**
지난 미션에서는 `innerHTML`을 사용해 구현했지만, 이번에는 `createElement`와 `appendChild` 같은 인스턴스 메서드를 활용하여 `DOM` 요소를 하나씩 추가하는 방식으로 변경했습니다. 그런데 이 과정에서 각 컴포넌트의 렌더링 함수가 길어지고, 코드 가독성이 떨어지는 문제가 발생했습니다. 보다 가독성 좋은 코드 구조를 만들 방법에 대한 피드백을 받고 싶습니다. 혹시 리팩토링을 통해 개선할 수 있는 더 좋은 방식이 있다면 추천해주시면 감사하겠습니다!

**✅ `ButtonGroup` 컴포넌트화 하는 것에 대하여**
위에서 말씀드린 `ButtonGroup`을 컴포넌트화하는 것에 대하여 마르코의 의견이 궁금합니다! 매개변수가 많아질수록 지나치게 많은 책임을 가지게 된다고 생각해서 오히려 더 복잡해질 것 같다는 생각이 들어요. 과연 컴포넌트에 대해 어디까지 역할을 부여해야 하는게 맞을까요?

**✅ `CSS` 없는 파일들에 대한 폴더 여부**
`CSS`를 각 컴포넌트와 동일한 폴더 하위에 위치시켜 좀 더 응집성을 높이도록 했습니다. 근데 이 과정에서 `CSS`가 존재하지 않는 파일들도 있었습니다. 하지만 이 과정에서 추후 `CSS`가 생성될 가능성을 염두에 두고, 통일성 유지를 위해 모두 폴더 안으로 넣어두었습니다. 예를 들면, `restaurantFormSection` 내부에 있는 컴포넌트들이나, `common`에 있는 `selectBox` 같은 것들이 있습니다. 하지만 너무 폴더 `Depth`가 깊어져서 오히려 보기가 불편해진 상황입니다. 이 상황에서 마르코라면 어떻게 파일들을 관리하실지 궁금합니다! 

**✅ 컴포넌트 추상화 수준과 접근성의 균형 고민**
`restaurantFormSection` 내부에는 총 6개의 컴포넌트가 있고, 가장 상위 부모 요소인 `restaurantForm` 컴포넌트와 그 안에 포함된 5개의 입력 폼(`categorySelect`, `descriptionInput`, `distanceSelect`, `linkInput`, `nameInput`)으로 구성되어 있습니다. 

여기서 추상화 레벨을 맞추기 위해서 `categorySelect`와 `distanceSelect`를 따로 분리했는데, 원래는 `restaurantForm` 내부에서 `selectBox`(`common`에 존재하는 컴포넌트)를 직접 선언해서 사용할 계획이었습니다. 하지만,`restaurantForm` 내부에서 각 컴포넌트를 일관된 방식으로 호출하기 위해 `categorySelect`와 `distanceSelect`를 별도로 만들게 되었습니다. 
아래는 해당 코드의 일부입니다.

```js
const $categoryFormItem = new CategorySelect().render();
const $nameFormItem = new NameInput().render();
const $distanceFormItem = new DistanceSelect().render();
const $descriptionFormItem = new DescriptionInput().render();
const $linkFormItem = new LinkInput().render();
```

문제는 이렇게 `categorySelect`와 `distanceSelect`를 감싸는 컴포넌트를 추가하면서 단순히 한 겹 더 감싼 느낌이 들어서 고민이 되었습니다. 물론 `restaurantForm` 내부에서 보면 구조적으로 통일감이 있고, 선언형 프로그래밍 패러다임에도 잘 맞아 가독성이 좋아보입니다. 하지만 내부를 들여다보면, 굳이 한 단계 더 거쳐야 하는 구조가 되어버려 불필요한 경로가 추가되었다는 생각이 들었어요. 

즉, `추상화 레벨을 맞추는 것`과 `접근성 높이는 것` 사이에서 균형을 맞추는 것이 고민입니다. 이 두 가지를 동시에 만족시킬 방법이 있을까요?

**✅ Cypress와 E2E 테스트에 대한 고민**

이번에 `Cypress`를 처음 사용해봤는데, 생각보다 `Jest`랑 달라서 고민이 있어서 질문 남깁니다.
`Cypress`에서 `Given-When-Then` 패턴처럼 가독성을 높이는 작성법이 있을까요?
그리고 `E2E` 테스트의 범위를 어디까지 설정해야 할지도 고민입니다. 
현재는 버튼 클릭부터 추가 확인까지 테스트했는데, 음식점을 여러 개 제출하는 경우까지 모두 테스트해야 할지 고민이 됩니다. 
그렇게 되면 테스트 내용이 너무 많이 길어져서 이것도 어떻게 코드를 효율적으로 짤 수 있을지도 고민됩니다.
혹시 이에 대한 가이드라인이나 추천하는 접근 방식이 있을까요? 

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

sooyeoniya committed:
docs: README.md 기능 요구사항 및 구현 목록 작성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(Header): Header 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 이미지 폴더 위치 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(RestaurantListItem): RestaurantListItem 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(RestaurantList): RestaurantList 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(SelectBox): SelectBox 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(BottomSheet): BottomSheet 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: html, css 파일 위치 이동 및 html 태그 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(Button): Button 컴포넌트 생성 밎 BottomSheet에 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(DescriptionInput): DescriptionInput 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(RestaurantNameInput): RestaurantNameInput 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(LinkInput): LinkInput 컴포넌트 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(BottomSheet): BottomSheet 내부에 분리된 컴포넌트 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(main): 전체 레이아웃 구성

- 구현한 내부 컴포넌트 가져와서 레이아웃 그리기
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 음식점 리스트 처음에 빈 값으로 초기화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(Header): 새로운 음식점 추가 버튼 클릭 시 모달 창 열기

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(BottomSheet): 취소하기 버튼 및 백드롭 클릭 시 모달 닫기

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(SelectBox): label 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(BottomSheet): 새로운 음식점 추가 기능 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
feat(constants): 카테고리 및 거리 배열 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
docs: README.md 리팩토링 및 테스트 목록 작성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(Button): 버튼 스타일 variant 분기 처리 리팩토링

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(BottomSheet): BottomSheet 컴포넌트 분리 및 SelectBox 폴더 위치 수정

- BottomSheet의 틀과 내부 폼 데이터를 각각 BottomSheetBase와 RestaurantForm으로 분리하여 폼을 주입하도록 함

- SelectBox는 common 폴더로 옮김

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(BottomSheet): 카테고리 및 거리 선택 박스 컴포넌트 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 사용하지 않는 import 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(Button): Button 컴포넌트 common 폴더로 경로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(RestaurantListSection): 레스토랑 리스트 및 아이템 폴더 경로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: BottomSheetBase 경로 common 폴더로 옮김

- BottomSheetBase는 공통으로 사용될 수 있는 바텀시트 틀이라고 판단하여 common 폴더로 옮김

- 그 과정에서 BottomSheetBase를 사용하는 RestaurantForm 관련 컴포넌트도 RestaurantFormSection 으로 폴더 경로 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(CSS): 응집성 향상을 위한 CSS 파일 분리 및 이동

- 각 컴포넌트 폴더에 위치하도록 함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 이벤트 타입 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(App): 루트 JS 파일을 App으로 변경

- main보다 App이라는 명칭을 사용해 파스칼 케이스로 컴포넌트 느낌을 주어서 함수로 묶어서 실행하도록 함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 각 label 명 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
fix(SelectBox): distance label textContent 수정

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
test: 새로운 음식점 추가 플로우 E2E 테스트 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
test: 바텀 시트 닫는 기능 E2E 테스트 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
docs: README.md 컴포넌트 폴더 구조 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 버튼 텍스트 상수화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: 전체 클래스 구조로 변경

- 함수 내부가 길어졌기 때문
- 렌더링과 이벤트 핸들링 로직 분리 및 해당 메서드들 재사용하기 위함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor(constants): 버튼 타입 상수화 및 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: 이벤트 핸들러 함수 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: 사용하지 않는 import 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
fix(BottomSheetBase): form 태그 중복 삭제

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
refactor: App에서 상태관리 되도록 구조 변경

- App에서 상태를 관리하고, 이를 하위 컴포넌트로 전달하여 중앙에서만 상태관리하도록 수정
- 2단계 작업에 대한 확장성을 고려하여 상태 변경을 일관된 방식으로 관리해줌

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
test: 필수 정보 입력하지 않았을 경우 음식점 추가되지 않는 플로우 테스트

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
chore: RestaurantNameInput 앞에 Restaurant 제거

- 다른 입력 박스들과 통일해주기 위함

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

sooyeoniya committed:
build: gh-pages 배포 경로 수정

sooyeoniya committed:
fix: 이미지 경로 수정

- public에서부터 상대경로로 설정

wonsss commented on README.md:
```diff
@@ -0,0 +1,83 @@
+├── 📂 header
+│ ├── Header.js
+│ └── header.css
+├── 📂 restaurantFormSection
```
폴더 네이밍 컨벤션으로 camelCase를 택하셨군요.
폴더 네이밍 컨벤션은 일반적으로 kebab-case나 PascalCase를 사용하는 것 같고, camelCase를 사용하는 경우는 잘 못 본 거 같아요.
물론 이건 팀 컨벤션으로 정하면 되는 거라서, 속한 팀에서 정한 컨벤션에 따라 일관되게만 하면 괜찮을 것 같아요.

기준 잡기 애매하시면 참고로 네이밍 컨벤션 아티클 하나 추천드려요.

<img width="875" alt="image" src="https://github.com/user-attachments/assets/941bbddd-d929-4540-a836-7675b09f2163" />

> https://www.newtonflash.com/technology/front-end-naming-conventions/

sooyeoniya replied:
저희는 camelCase로 적용했는데, 찾아보니 일반적으로 JS 프로젝트에서는 kebab-case가 많이 사용되고, 컴포넌트 폴더에서는 PascalCase도 많이 사용되는 것 같네요. 좋은 정보 감사합니다! https://github.com/woowacourse/javascript-lunch/commit/6d80a27ed874c8f5c8c9d96d3c5b8e899f1e4132

wonsss commented on cypress/e2e/AddNewRestaurant.cy.js:
```diff
@@ -0,0 +1,83 @@
+    cy.get('input[name="link"]')
+      .should("be.visible")
+      .type("링크 테스트용 텍스트")
+      .should("have.value", "링크 테스트용 텍스트");
```
form 에 값을 입력(type, select 메서드 등)하는 코드네요. 
이는 given/when/then 3단계에서, when 블록에 해당할 것 같아요.
여기에 함께 있는 should 메서드는 then 블록으로 분리해서 실행하는 게 테스트 흐름이 더 자연스러울 것 같아요. 테스트코드도 코드이기 때문에 다른 코드들과 동일하게 가독성과 유지보수성을 고려해주시면 좋아요.

그리고 이러한 form 입력하기가 밑에서도 반복되기 때문에 이를 다음과 같은 인터페이스의 함수로 분리할 수 있을 것 같아요. 

`fillForm({ category, name, distance, description, link  })`

sooyeoniya replied:
재사용 로직에 대해서는 내부 헬퍼 함수로 분리해줬습니다. 브라우저 환경 초기화해주는 로직에 대해서도 커스텀 커맨드로 정의해서 재사용 해줬습니다! 

아래 말씀해주신 테스트 코드 관련해서도 모두 반영해두었습니다. https://github.com/woowacourse/javascript-lunch/pull/183/commits/b27f8727a16393662099442542d1c6ffb53b93c0

확실히 코드가 더 깔끔해진 것 같다는 느낌이 듭니다 :)

wonsss commented on cypress/e2e/AddNewRestaurant.cy.js:
```diff
@@ -0,0 +1,83 @@
+  });
+
+  const openFormAndCheckLabels = () => {
+    cy.get(".gnb__button").click();
```
openFormAndCheckLabels 함수가 두 가지 역할을 함께 하네요.
`cy.get(".gnb__button").click();` 만 openForm 으로 별도 분리하고, 나머지는 checkFormLabels함수로 두는 게 명확해보여요.

wonsss commented on cypress/e2e/AddNewRestaurant.cy.js:
```diff
@@ -0,0 +1,83 @@
+      .should("exist")
+      .and("be.visible");
+    cy.contains("label", "설명").should("exist").and("be.visible");
+    cy.contains("label", "참고 링크").should("exist").and("be.visible");
```
규칙성이 있어보여 forEach 같은 메서드로 반복문을 만들면, 가독성이 더 좋아질 것 같아요.

wonsss commented on src/App.js:
```diff
@@ -0,0 +1,42 @@
+import BottomSheetBase from "./components/common/bottomSheetBase/BottomSheetBase.js";
+import RestaurantForm from "./components/restaurantFormSection/restaurantForm/RestaurantForm.js";
+
+export default class App {
```
App 혼자 너무 많은 역할을 담당한다고 볼 수도 있어요. 
상태 관리와 UI 초기화, 비즈니스 로직, 이벤트 핸들링 등을 담당하고 있네요.

상태 관리와 비즈니스 로직은 각각 별도의 모듈로 분리해보면 좋을 거 같습니다.
`this.restaurantList = [];` 처럼 App이 직접 상태를 관리하게 되면, 나중에 상태들이 늘어났을 때 유지보수가 어려워요.

그리고 비즈니스 로직도 분리해볼 수 있어요.
비즈니스 로직이란, 앱이 해결하려는 문제의 본질이라고 생각해요. 여기서는 음식점 추가하고 목록 관리하는 게 해결하려는 문제이고, 결국 비즈니스 로직은 음식점 데이터를 관리하는 모든 로직이에요.

UI가 변경되더라도 비즈니스 로직이 영향을 받지 않도록 분리해보시는 연습을 계속해보시면 도움이 되실 거예요.

wonsss replied:
App 쪽에 대한 리뷰는 이번 미션 목표보다는 이전 미션 목표에 가까운 것 같기도 하네요.
우선 다른 리뷰들 반영해보시고 여력이 되시면, 이 부분 리뷰들은 반영해주시거나, 참고만 해주셔도 괜찮을 것 같아요.

sooyeoniya replied:
좋은 지적 감사합니다! 1단계 미션 수행 당시에는 빠르게 구현하고 피드백 받고 싶어서 제대로 확인 못한 부분이 꽤 많은 것 같네요 :-)

2단계 미션 작업 진행할 때 기능 추가할수록 더욱 스파게티 코드가 될 것 같다는 생각이 들었습니다. 따라서, 2단계 미션 작업을 위해서는 분리해야하는 부분이라고 생각했습니다!

wonsss replied:
멋집니다 니야 👍

wonsss commented on src/App.js:
```diff
@@ -0,0 +1,42 @@
+
+  #addList = (newRestaurantInfo) => {
+    this.restaurantList = [...this.restaurantList, newRestaurantInfo];
+    this.#renderRestaurantList();
```
비즈니스 로직과 UI 로직이 섞여 있네요.  분리해보면 좋을 것 같아요.

wonsss commented on src/App.js:
```diff
@@ -0,0 +1,42 @@
+
+export default class App {
+  constructor() {
+    this.restaurantList = [];
```
상태를 App이 아닌 별도 모듈에서 관리해보시는 것도 고민해볼 수 있어요. 
지금처럼 간단하고 작은 프로젝트의 경우 괜찮을 수 있어요.
다만, 우선 App 역할이 많고, 비즈니스 로직이 많아지면 App에서 직접 상태를 관리할 경우 코드가 복잡해져 확장성이 떨어져요. 그리고 상태를 여러 컴포넌트에서 공유하기가 어려워요. 여러 컴포넌트에서 상태를 사용해야 할 때, App이 직접 상태를 관리하면 상태 전달이 필요해져 복잡도가 증가할 수 있어요.

wonsss commented on src/components/restaurantListSection/restaurantList/RestaurantList.js:
```diff
@@ -0,0 +1,23 @@
+    );
+
+    return $listSection;
+  }
```
> ✅ 각 컴포넌트의 렌더링 방식 개선 고민
지난 미션에서는 innerHTML을 사용해 구현했지만, 이번에는 createElement와 appendChild 같은 인스턴스 메서드를 활용하여 DOM 요소를 하나씩 추가하는 방식으로 변경했습니다. 그런데 이 과정에서 각 컴포넌트의 렌더링 함수가 길어지고, 코드 가독성이 떨어지는 문제가 발생했습니다. 보다 가독성 좋은 코드 구조를 만들 방법에 대한 피드백을 받고 싶습니다. 혹시 리팩토링을 통해 개선할 수 있는 더 좋은 방식이 있다면 추천해주시면 감사하겠습니다!

createElement와 appendChild 메서드 사용하여 안전하게 잘 만드셨네요. 다만 느끼신 것처럼 간단한 컴포넌트인데도 가독성이 떨어지는 문제가 있죠. DOM 조작은 성능에 큰 영향을 미치며, 비효율적인 방식으로 다루면 렌더링 속도 저하와 메모리 누수가 발생할 수 있어요.

이를 개선하기 위한 방법 중 하나로 `<template>` 태그도 사용해볼 수도 있어요.

꼭 이 방법을 사용하진 않아도 되고, 해당 문제의식에서 니야만의 더 좋은 방법을 고민해보시는 연습을 해보시면 도움이 되실거예요.

> 참고
"HTML template 요소는 페이지를 불러온 순간 즉시 그려지지는 않지만, 이후 JavaScript를 사용해 인스턴스를 생성할 수 있는 HTML 코드를 담을 방법을 제공합니다."
https://developer.mozilla.org/ko/docs/Web/HTML/Element/template
https://blog.webdevsimplified.com/2020-06/template-tag/

sooyeoniya replied:
덕분에 template 태그에 대해 공부하고 갑니다 !! template 태그를 쓰면 DOM을 직접 조작해야 활성화가 되니까 textContent 같은 메서드랑 함께 잘 활용하면 동적으로 html을 안전하게 잘 삽입할 수 있겠네용

wonsss commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+    const link = document.querySelector("#link");
+
+    return { category, name, distance, description, link };
+  }
```
#getFormQuery()는 document.querySelector()를 반복 호출하여 불필요한 DOM 접근을 하게 되네요.
이를 클래스 필드로 저장해두고 클래스 내부에서 재사용하게 되면 성능 개선도 되고 더 직관적일 수 있어요.

```js
  constructor(addList) {
    this.addList = addList;
    this.formElements = {
      category: new CategorySelect().render(),
      name: new NameInput().render(),
      distance: new DistanceSelect().render(),
      description: new DescriptionInput().render(),
      link: new LinkInput().render(),
    };
  }

```

sooyeoniya replied:
해당 부분에 대해서 고쳐봤습니다! 확실히 querySelector로 한 번 더 DOM에 접근하는게 불필요한 접근인 것 같다는 생각이 들었어요. 말씀대로 필드에 객체로 저장해두고 접근하도록 수정해주었습니다 👍🏻  https://github.com/woowacourse/javascript-lunch/pull/183/commits/2cca0ef6078ae14b18024efda7d19884c3ab2134

wonsss commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+        return acc;
+      },
+      {}
+    );
```
handleSubmit이 많은 역할을 담당하고 있는데, 이처럼 데이터를 가져오는 부분은 #getFormData() 같은 함수로 분리해볼 수 있을 것 같아요.

sooyeoniya replied:
함수 역할 분리해서 좀 더 가독성 있게 해봤습니다! 확실히 분리하니까 해당 로직이 어떤 것을 의미하는지 선언적으로 알 수 있어서 보기 좋은 것 같습니다. https://github.com/woowacourse/javascript-lunch/pull/183/commits/5bd9ae48a035ed75f15184cf12b565c2b0463c62

wonsss commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+  #closeModal() {
+    const $modal = document.querySelector(".modal");
+    $modal.classList.remove("modal--open");
+  }
```
#closeModal이 #handleCancelButtonClick 의 코드가 동일하네요. 중복 제거할 수 있겠네요.

```
  #handleCancelButtonClick() {
    const $modal = document.querySelector(".modal");
    $modal.classList.remove("modal--open");
  }
```

sooyeoniya replied:
중복 코드 제거했습니다 👍🏻 정말 눈이 침침한가봐요.. 구현하다보니 같은 로직을 두번이나 작성해두다니 ㅠㅠ https://github.com/woowacourse/javascript-lunch/pull/183/commits/d021b575fba8893fc77927ea77055f49b6b49432

wonsss commented on src/components/restaurantFormSection/categorySelect/CategorySelect.js:
```diff
@@ -0,0 +1,13 @@
+
+    return $categorySelect;
+  }
+}
```
> ✅ 컴포넌트 추상화 수준과 접근성의 균형 고민
restaurantFormSection 내부에는 총 6개의 컴포넌트가 있고, 가장 상위 부모 요소인 restaurantForm 컴포넌트와 그 안에 포함된 5개의 입력 폼(categorySelect, descriptionInput, distanceSelect, linkInput, nameInput)으로 구성되어 있습니다.
여기서 추상화 레벨을 맞추기 위해서 categorySelect와 distanceSelect를 따로 분리했는데, 원래는 restaurantForm 내부에서 selectBox(common에 존재하는 컴포넌트)를 직접 선언해서 사용할 계획이었습니다. 하지만,restaurantForm 내부에서 각 컴포넌트를 일관된 방식으로 호출하기 위해 categorySelect와 distanceSelect를 별도로 만들게 되었습니다.
아래는 해당 코드의 일부입니다.
const $categoryFormItem = new CategorySelect().render();
const $nameFormItem = new NameInput().render();
const $distanceFormItem = new DistanceSelect().render();
const $descriptionFormItem = new DescriptionInput().render();
const $linkFormItem = new LinkInput().render();
문제는 이렇게 categorySelect와 distanceSelect를 감싸는 컴포넌트를 추가하면서 단순히 한 겹 더 감싼 느낌이 들어서 고민이 되었습니다. 물론 restaurantForm 내부에서 보면 구조적으로 통일감이 있고, 선언형 프로그래밍 패러다임에도 잘 맞아 가독성이 좋아보입니다. 하지만 내부를 들여다보면, 굳이 한 단계 더 거쳐야 하는 구조가 되어버려 불필요한 경로가 추가되었다는 생각이 들었어요.
즉, 추상화 레벨을 맞추는 것과 접근성 높이는 것 사이에서 균형을 맞추는 것이 고민입니다. 이 두 가지를 동시에 만족시킬 방법이 있을까요?

좋은 고민이네요! 선언형 방식 유지 및 확장 가능성 고려라는 측면에서 현재 방식도 적절하다고 생각해요.
그렇다고 해서 반드시 이렇게 추상화 레벨을 맞추기 위해 단순히 감싸기만 하는 Wrapper 컴포넌트를 만들어야 하는 건 아니라서 참고차 말씀드려요.
현재 CategorySelect와 DistanceSelect는 단순히 라벨과 옵션을 제공하는 SelectBox의 wrapper 역할만 하고 있어요. 그래서 사실 이러한 Wrapper 컴포넌트를 만들지 않고, 사용처에서 SelectBox를 바로 호출하더라도 가독성은 괜찮다고 생각해요. 현재 자바스크립트로 만들고 있어서 코드가 좀 길어지는 느낌이 있긴 하지만 이 코드가 jsx라 생각한다면 이 정도로 직관적인 컴포넌트이기 때문이에요.
```jsx
<SelectBox label="category" options="CATEGORY" />
```

추상화 레벨을 항상 동일하게 맞추기는 어렵다고 생각해요.
일관성이 중요하지만, 불필요한 추상화는 오히려 유지보수를 어렵게 만들 수도 있어요. 
실제 기능과 필요성에 따라 추상화 수준을 조정하는 것이 더 중요할 수 있어요.
즉, 추상화 레벨을 맞추는 것이 목적이 되는 게 아니라, 유지보수성과 가독성을 높이는 것을 목표로 하다보면 적절한 추상화를 사용하게 되는 것 같아요.

만약 확장 가능성이 있다면 지금 작성하신 것처럼 각각 독립적인 컴포넌트로 유지하는 게 더 적절해요.
예를 들어, CategorySelect 컴포넌트와 DistanceSelect 컴포넌트 각각 다른 API를 호출하여 응답값으로 옵션을 동적으로 구성해야 한다거나, 각자 다른 비활성화 정책이 존재해야 하는 등의 경우에서요.

너무 좋은 접근 방식이네요. 👍  답변이 되었으면 좋겠네요.

sooyeoniya replied:
꼼꼼한 답변 정말 감사합니다! 😃 유지보수성 측면에서 봤을 때 추가 로직이 붙을 경우에는 미리 분리해두는 것도 좋은 접근법인 것 같네요! 하지만, 사실 본 프로젝트에서 봤을 때는 추가 로직이 따로 붙지는 않을 것 같아요. 그래서 불필요한 컴포넌트 두개 삭제해주었습니다 ! https://github.com/woowacourse/javascript-lunch/pull/183/commits/77099da606156126c6ce2eb73209b753e17ffb1c

wonsss replied:
👍

wonsss commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+      type: "submit",
+      text: BUTTON_TEXTS.add,
+      action: BUTTON_TYPES.add,
+    }).render();
```
> ✅ ButtonGroup 컴포넌트화 하는 것에 대하여
위에서 말씀드린 ButtonGroup을 컴포넌트화하는 것에 대하여 마르코의 의견이 궁금합니다! 매개변수가 많아질수록 지나치게 많은 책임을 가지게 된다고 생각해서 오히려 더 복잡해질 것 같다는 생각이 들어요. 과연 컴포넌트에 대해 어디까지 역할을 부여해야 하는게 맞을까요?


컴포넌트도 하나의 명확한 역할을 가지도록 하고 과도한 책임을 가지지 않도록 해야 해요.
만약 너무 많은 기능을 포함하면 재사용성이 낮아지고, 유지보수가 어려운 복잡한 컴포넌트가 될 수도 있어요.

또한, 컴포넌트가 여러 곳에서 재사용될 가능성이 있다면 그 자체로 적절한 응집도의 컴포넌트라고 볼 수 있어요.

그렇기 때문에 ButtonGroup을 컴포넌트화할 때는 어떤 역할을 수행할지 명확하게 정의해보면 좋아요

- 도메인에 특화된 버튼 그룹
   - 취소기능과 추가기능을 각각 가진 두 버튼의 그룹 
   - RestaurantForm과 같은 특정 UI에서만 사용됨
   - 도메인에 강하게 연관되므로 재사용성이 낮음
- 일반적인 버튼 그룹
   - 주기능과 보조기능을 각각 가진 두 버튼의 그룹 
   - 조금 더 일반적이라서 재사용 가능성 높아짐  
   - 이벤트 핸들러를 외부에서 주입하여 유연하게 사용 가능
- 단순히 버튼을 배치하는 역할만 수행 
   - 버튼 개별 로직은 Button에서 처리하고 ButtonGroup은 단순 배치만 담당
   - 그런데 버튼 UI가 고정된 형태라면 불필요할 수 있음

이외에도 다른 역할을 정의해볼 수도 있고, 상황에 따라 적절성이 다르므로 이들 중 무엇이 가장 좋다고 말하기도 어려워요. 

말씀하신 것처럼 우선 YAGNI 원칙대로 단순하게 개발하고 공통 패턴이 반복될 때 추상화하는 것도 적절하다고 생각합니다. 또한, 재사용성을 고려하되 필요 이상으로 일반화하지 않아도 된다고 생각해요.

sooyeoniya replied:
확실히 로직이 완전 다르고, 단순히 컴포넌트 스타일만 재사용하기 위해 ButtonGroup 컴포넌트를 따로 생성하는 것은 너무 과도하게 추상화하는 느낌도 드네요! 우선 YAGNI 원칙대로 지금 그대로 가보겠습니다 ㅎㅎ 2단계에서 분리할 필요성을 느끼면 그때 적용해보겠습니다! 정성스러운 답변 감사해요 😊

wonsss commented on src/components/common/selectBox/SelectBox.js:
> ✅ CSS 없는 파일들에 대한 폴더 여부
CSS를 각 컴포넌트와 동일한 폴더 하위에 위치시켜 좀 더 응집성을 높이도록 했습니다. 근데 이 과정에서 CSS가 존재하지 않는 파일들도 있었습니다. 하지만 이 과정에서 추후 CSS가 생성될 가능성을 염두에 두고, 통일성 유지를 위해 모두 폴더 안으로 넣어두었습니다. 예를 들면, restaurantFormSection 내부에 있는 컴포넌트들이나, common에 있는 selectBox 같은 것들이 있습니다. 하지만 너무 폴더 Depth가 깊어져서 오히려 보기가 불편해진 상황입니다. 이 상황에서 마르코라면 어떻게 파일들을 관리하실지 궁금합니다!

현재 컴포넌트별 폴더를 유지하는 방식이 일관성이 있고 향후 css나 테스트 코드 등이 추가될 경우를 생각했을 때, colocation 측면에서 좋다고 생각해요!

wonsss commented on cypress/e2e/AddNewRestaurant.cy.js:
```diff
@@ -0,0 +1,83 @@
+describe("새로운 음식점 추가 플로우 테스트", () => {
```
> ✅ Cypress와 E2E 테스트에 대한 고민
이번에 Cypress를 처음 사용해봤는데, 생각보다 Jest랑 달라서 고민이 있어서 질문 남깁니다.
Cypress에서 Given-When-Then 패턴처럼 가독성을 높이는 작성법이 있을까요?
그리고 E2E 테스트의 범위를 어디까지 설정해야 할지도 고민입니다.
현재는 버튼 클릭부터 추가 확인까지 테스트했는데, 음식점을 여러 개 제출하는 경우까지 모두 테스트해야 할지 고민이 됩니다.
그렇게 되면 테스트 내용이 너무 많이 길어져서 이것도 어떻게 코드를 효율적으로 짤 수 있을지도 고민됩니다.
혹시 이에 대한 가이드라인이나 추천하는 접근 방식이 있을까요?

하나의 질문에 여러 개가 함께 포함되어 있네요. 그만큼 E2E 테스트가 쉽지 않은 것 같아요. ㅎㅎ 

> Cypress에서 Given-When-Then 패턴처럼 가독성을 높이는 작성법

Cypress에서도 동일하게 Given-When-Then 패턴을 사용할 수 있어요. 
가독성을 높이는 작성법은 다른 코멘트에서도 말했듯이, 테스트 흐름이 명확하게 보일 수 있도록 패턴대로 분리해보시는 게 좋아요.
그리고 테스트코드도 코드이므로 반복되는 로직은 헬퍼 함수나 [Cypress custom command](https://docs.cypress.io/api/cypress-api/custom-commands) 기능으로 분리하여 가독성을 높일 수 있어요.

> E2E 테스트의 범위를 어디까지 설정해야 할지

여력이 된다면 꼼꼼하게 많이 작성하는 게 좋죠.
그러나 E2E 테스트는 작성 및 유지하는 데에도 비용이 많이 드는 편이기 때문에,  핵심 사용자 플로우 중심으로 작성하는 게 좋아요. 새로운 음식점 추가 그리고 지금은 없지만 예를 들어 음식점 삭제, 수정 등의 플로우는 E2E로 하는 게 좋고, 세부적인 유효성 검사 등의 로직은 유닛 테스트에서 처리할 수 있어요.

> 음식점을 여러 개 제출하는 경우까지 모두 테스트해야 할지 고민

시나리오를 생각했을 때, 이것을 테스트할 필요성이 있는지 고민해보시고 혹시 걱정이 된다면 테스트를 하는 게 맞아요.
예를 들어, 음식점을 여러 개 제출했을 때, 이전에 입력한 음식점이 목록에서 사라지거나 다르게 배치되는 등의 문제가 우려된다면 사용자 경험에 문제가 생기겠죠. 이 경우 음식점 한 개뿐만 아니라 여러 개 추가했을 때에도, 목록이 정상적으로 업데이트되는지 검증할 필요가 있다고 생각해요. 

> 참고
https://developers.kakaomobility.com/docs/techblogs/e2e-testing

sooyeoniya replied:
본 기능 자체가 음식점 목록에 새로운 음식점을 하나씩 추가해서 그 목록 리스트를 보여주는 것이기 때문에, 사용자 입장에서 보면, 새로운 음식점을 추가했을 때 음식점이 목록에 계속 누적되어 보여야 한다고 생각했습니다. 이런 관점에서 본다면 시나리오 상 여러 개의 음식점을 등록하는 경우도 필요하다고 생각했습니다! https://github.com/woowacourse/javascript-lunch/pull/183/commits/dec333cf08cbc5ad475f5f7b333dedc0311ef00b

앞선 테스트 코드 관련 리뷰 답장에서도 보셨다시피 Given-When-Then 패턴과 헬퍼함수, 커스텀 커맨드도 활용해서 코드를 깔끔하게 해봤습니다 :) Given-When-Then 패턴이 정확하게 구분된 건지는 잘 모르겠지만, 제가 이해한 흐름대로 최대한 분류해봤어요!

wonsss replied:
"여러 음식점이 추가되는 경우에 대한 테스트 케이스" 추가 잘해주셨네요 👍

wonsss commented on src/components/common/bottomSheetBase/BottomSheetBase.js:
```diff
@@ -0,0 +1,43 @@
+import { EVENT_TYPES } from "../../../constants/constants.js";
+import "./bottomSheetBase.css";
+
+export default class BottomSheetBase {
```
현재 바텀시트 컴포넌트는 바텀시트를 열고 닫는 기능을 직접 제공하지 않고 있네요.
그래서 다른 컴포넌트들에서 바텀시트를 열고 닫기 위해 `modal--open` 클래스를 직접 DOM에서 조작하는 명령형 방식을 쓰고 있어요. 그런데 이는 나중에 가독성이나 유지보수가 어려운 방식이라고 볼 수 있어요.
open() 및 close() 메서드를 추가하여 컴포넌트 내에서 바텀시트를 열고 닫을 수 있도록 개선해보면 좋을 것 같아요.

그리고 Header나 Form 내에서 직접 DOM 접근하여 클래스 통해 바텀시트 열고 닫았던 로직은 컴포넌트에서 콜백함수를 인자로 받아서 처리해주는 게 적절해보여요. Header나 Form 입장에서는 바텀시트가 존재한다는 것을 몰라야 해요.

sooyeoniya replied:
바텀 시트 컴포넌트에 대한 기능이 다 외부에 있었네요.. 😅 내부로 옮겨주었습니다 ! https://github.com/woowacourse/javascript-lunch/pull/183/commits/c09b12b1779a49a047648bfb2e0cb3fd68108814

아래 RestaurantForm 관련 바텀 시트 로직도 전부 수정해줬습니다!

wonsss replied:
잘 수정해주셨네요 👍  이처럼 컴포넌트가 응집성 있게 작성해주시면 좋아요.

wonsss commented on src/components/restaurantFormSection/restaurantForm/RestaurantForm.js:
```diff
@@ -0,0 +1,107 @@
+  BUTTON_TYPES,
+} from "../../../constants/constants.js";
+
+export default class RestaurantForm {
+  constructor(addList) {
+    this.addList = addList;
+  }
```
RestaurantForm 컴포넌트도 매개변수로 콜백함수를 좀 더 범용성있게 구성하면 좋을 것 같아요.
- 폼 제출 시 실행할 콜백 : onSubmit (addList라는 이름보다 form 입장에서 역할)
- 취소 버튼 클릭 시 실행할 콜백: onCancel (부모 컴포넌트에서 인자로 'closeModal()' 등이 전달될 수 있음)

sooyeoniya replied:
폼 관련 로직이 한 눈에 제대로 안보였는데, 리팩토링하니까 훨씬 가독성도 좋고 갖다 쓰기 편하게 되어서 좋은 것 같습니다 😃 아래 작성해주신 예시 인터페이스 코드 덕분에 수월하게 리팩토링할 수 있었습니다! 모달 닫는 로직을 전부 외부에서 주입해주도록 바꿨어요!
https://github.com/woowacourse/javascript-lunch/pull/183/commits/46e61cfa204ca36c98b41a9dd2be8d7939c838ad

wonsss replied:
가독성있게 잘 수정해주셨네요!

wonsss commented on src/App.js:
```diff
@@ -0,0 +1,42 @@
+    this.$main.appendChild(
+      new BottomSheetBase({
+        title: "새로운 음식점",
+        $children: new RestaurantForm(this.#addList).render(),
```
위 코멘트에서 RestaurantForm 개선된 인터페이스 호출 예시는 다음과 같이 생각해봤어요. 설명을 쉽게 하고 싶어서 예시를 전달 드렸어요. 이건 참고만 하시고, 수정하셔도 돼요.
```js
      $children: new RestaurantForm({
        onSubmit: (data) => {
             this.#addList(data);
             this.bottomSheet.close();
        }, 
        onCancel: () => this.bottomSheet.close(), 
      }).render(),
```

wonsss reviewed (approved):
안녕하세요. 니야 😄  이번 미션 리뷰를 맡은 마르코입니다!  🙇 
미션을 잘 진행해주셨네요!
UI를 컴포넌트 단위로 잘 분리해주셨고, 재사용 가능한 컴포넌트를 만들기 위해 적절한 추상화 정도에 대해 고민을 해주셨던 것 같아요. 
좋은 질문을 많이 주셔서 저도 덕분에 좋은 고민을 해볼 수 있었어요. 👍 
본문에서 주신 질문들에 대해서는 관련 맥락을 함께 논의해볼 수 있도록 해당 코드에 댓글로 답변을 남겨두었어요.
확인 후 재리뷰 요청주세요~! 감사합니다.

wonsss reviewed (changes requested):
안녕하세요. 니야 😄 이번 미션 리뷰를 맡은 마르코입니다! 🙇
미션을 잘 진행해주셨네요!
UI를 컴포넌트 단위로 잘 분리해주셨고, 재사용 가능한 컴포넌트를 만들기 위해 적절한 추상화 정도에 대해 고민을 해주셨던 것 같아요.
좋은 질문을 많이 주셔서 저도 덕분에 좋은 고민을 해볼 수 있었어요. 👍
본문에서 주신 질문들에 대해서는 관련 맥락을 함께 논의해볼 수 있도록 해당 코드에 댓글로 답변을 남겨두었어요.
확인 후 재리뷰 요청주세요~! 감사합니다.

sooyeoniya committed:
chore: 폴더 네이밍 camelCase -> kebab-case로 변경

- 폴더 네이밍 관련하여 가장 일반적인 컨벤션인 kebab-case로 변경함

sooyeoniya committed:
chore: E2E 테스트 로컬 주소 변경

sooyeoniya committed:
refactor(test): 테스트코드 리팩토링 및 테스트 환경 초기화 로직 Cypress 커스텀 커맨드로 분리

- AddNewRestaurant.cy.js 파일 내부 가독성 및 유지보수성 향상을 위한 리팩토링 진행
- 두 개의 테스트 파일에서 공통으로 사용되는 테스트 환경 초기화 로직을 Cypress의 커스텀 커맨드로 분리하여 적용

sooyeoniya committed:
test: 여러 음식점이 추가되는 경우에 대한 테스트 케이스 작성

- 이전 음식점 목록이 사라지지 않고 정상적으로 화면에 보이는지 확인하기 위한 시나리오 테스트
- 추가하기 버튼 클릭 로직 헬퍼함수로 분리하여 재사용

sooyeoniya committed:
refactor: appendChild -> append 로 변경

- append를 선택한 이유
1. 자식 노드 여러 개 추가 가능
2. Node 객체 외에 문자열도 추가 가능 -> 본 프로젝트에서는 Node 객체만 추가하고 있긴함
3. appendChild의 return 값을 굳이 사용하지 않음. append는 return 값이 없기 때문에 더 적합하다고 판단함
4. 코드가 훨씬 간결해지고 가독성 좋아짐

sooyeoniya committed:
refactor: 불필요한 DOM 반복 접근 최소화

- querySelector로 한 번 더 DOM에 반복 접근 이슈 수정
- 클래스 필드에 formElement 객체 저장하여, 각 요소를 초기화할 때 한 번만 DOM에 접근하도록 수정
- 성능 개선 및 코드 가독성 향상

sooyeoniya committed:
refactor: 모달 닫기 중복 로직 제거

- handleCancelButtonClick, closeModal 메서드 로직 중복 사용

sooyeoniya committed:
refactor(RestaurantForm): 입력 폼의 전체 값들 가져오는 로직 분리

- handleSubmit에서 분리할 수 있는 내부 로직이 존재함
- 입력 폼의 전체 값들을 가져오는 로직을 getFormData 라는 메서드로 분리함

sooyeoniya committed:
refactor(RestaurantForm): 매개변수 인터페이스 정의

- RestaurantForm 컴포넌트 매개변수의 콜백함수를 좀 더 범용성있게 수정
- onSubmit, onCancel 를 가진 인터페이스 정의하여 폼으로써의 역할을 명확히 하도록 함
- 모달 닫는 로직을 BottomSheetBase 컴포넌트로 책임을 완전히 위임 -> RestaurantForm, BottomSheetBase 역할 분리

sooyeoniya committed:
refactor(Header): 헤더의 모달 열기 로직 BottomSheetBase 로 이양

- BottomSheetBase 내부에서 바텀 시트가 열고 닫히는 기능을 맡도록 수정
- Header 에서는 버튼 클릭 시 onOpen이라는 콜백함수를 사용하고 외부에서 해당 콜백을 주입해줌
- 따라서, Header에서는 바텀시트의 존재 여부를 알 수 없음 -> 결합도 낮춤

sooyeoniya committed:
refactor: Observer 패턴으로 상태 관리 분리 및 음식점 관련 비즈니스 로직 분리

- App 에서 상태관리 부분과 비즈니스 로직을 각각 RestaurantStore, RestaurantService 로 분리하여 관리함
- RestaurantStore는 Observer 패턴 적용하여 상태 변경 시 목록 리스트 업데이트 되도록 해당 부분 구독

sooyeoniya committed:
refactor: RestaurantList 업데이트 로직 위치 변경

- 음식점 목록 업데이트 로직을 RestaurantList 내부로 이동
- RestaurantList의 생성자 함수에서 정적 DOM 요소들을 초기화하도록 옮김
- render 내부에서는 음식점 목록들(li 태그들)만 현재 음식점 목록 정보로 렌더링 되도록 함
- update 에서는 현재 음식점 목록 정보를 업데이트 하고 render를 호출해줌

sooyeoniya committed:
refactor(App): DI 적용 및 App 구조 리팩토링

- App 클래스 생성자에서 RestaurantStore와 RestaurantService를 직접 생성하지 않고, 의존성을 외부에서 주입받도록 수정
- App의 인스턴스화 로직을 index.js로 이동
- 초기 렌더링 로직을 render() 메서드로 명칭 변경 -> 렌더링 메서드 전체 통일
- RestaurantForm 인스턴스를 별도로 분리하여 BottomSheetBase에 전달
- App 에서 store와 service를 직접 관리하지 않음 -> App에서는 UI 로직만 담당

sooyeoniya committed:
refactor(App): UI 렌더링 로직 모듈화하여 가독성 및 유지보수성 개선

- renderHeader, renderMain 메서드로 분리하여 역할 명확히 지정
- render 함수들과 역할 분리하기 위해 updateRestaurantList 메서드 네이밍 변경
- onSubmit 핸들러 메서드 분리

sooyeoniya committed:
chore: 사용하지 않는 import 제거

sooyeoniya committed:
refactor(RestaurantForm): 확장성 측면에서 불필요한 추상화 제거

- SelectBox를 단순히 한 번만 더 감싸고 있는 CategorySelect, DistanceSelect 컴포넌트 제거
- 확장성을 고려해봤을 때 추가 로직이 붙을 가능성이 있는 경우 유지해도됨
- 하지만, 본 프로젝트에서는 더 이상 해당 부분들에 확장될 기능이 따로 없다고 판단하여 삭제함

sooyeoniya commented:
안녕하세요, 마르코! 니야입니다 🐢
리뷰 꼼꼼하게 해주셔서 정말 감사합니다! 덕분에 고민도 많이 해보고 정말 도움 많이 된 것 같아요 ㅎㅎ
아래는 리뷰 반영 요약표입니다. 참고하시면 커밋 내용을 더욱 쉽게 확인하실 수 있을 거예요! 😊
커밋 별 자세한 내용은 커밋 내역에 상세하게 적어두었습니다! 이것도 참고해주시면 감사하겠습니다 😁

## ✅ 리뷰 반영 요약표

내용 | 관련 코멘트 | 커밋
-- | -- | --
폴더 네이밍 컨벤션 | 폴더 네이밍 컨벤션 관련 코멘트 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/6d80a27ed874c8f5c8c9d96d3c5b8e899f1e4132
테스트 코드 리팩토링 | form 채우는 로직 리팩토링 두 가지 역할하는 함수 규칙성 있는 로직 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/b27f8727a16393662099442542d1c6ffb53b93c0
음식점 여러 개를 제출하는 경우 시나리오 테스트 | 테스트 코드 관련 질의응답 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/dec333cf08cbc5ad475f5f7b333dedc0311ef00b
불필요한 DOM 반복 접근 | 불필요한 DOM 반복 접근 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/2cca0ef6078ae14b18024efda7d19884c3ab2134
입력 폼 데이터 로직 분리 | handleSubmit 메서드 역할 분리 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/5bd9ae48a035ed75f15184cf12b565c2b0463c62
모달 닫기 중복 로직 제거 | 중복 코드 관련 코멘트 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/d021b575fba8893fc77927ea77055f49b6b49432
확장성 측면에서 불필요한 추상화 제거 | 추상화 레벨에 대한 코멘트 | https://github.com/woowacourse/javascript-lunch/commit/77099da606156126c6ce2eb73209b753e17ffb1c
바텀시트 내부에서 바텀시트 열고 닫는 역할 부여 | 바텀시트 컴포넌트 상태 관리 및 제어 방식 개선 제안 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/c09b12b1779a49a047648bfb2e0cb3fd68108814
RestaurantForm 콜백 함수 네이밍 및 인터페이스 개선 | RestaurantForm 컴포넌트 매개변수 구성 개선된 RestaurantForm 매개변수 인터페이스 예시 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/46e61cfa204ca36c98b41a9dd2be8d7939c838ad
상태 관리 및 비즈니스 로직과 UI 로직 분리하여 내부 구조 개편 | App 역할 과다 문제 비즈니스 로직과 UI 로직 혼합 상태 관리 분리 필요 | https://github.com/woowacourse/javascript-lunch/pull/183/commits/8acaf714790e47cde280e297e46185eecae0dd2e https://github.com/woowacourse/javascript-lunch/pull/183/commits/c0177ca14216a37a2f4487019eb13887cc95a8d2 https://github.com/woowacourse/javascript-lunch/pull/183/commits/3d3032d32b11b0c938ab415b9e789f65c2f972e5 https://github.com/woowacourse/javascript-lunch/pull/183/commits/9b1ce3dadd2921ca133b9bebb8894eabcad988f4

wonsss reviewed (approved):
안녕하세요! 니야~! 1단계 리뷰 잘 반영해주셨네요. 👍 
코멘트마다 니야가 느낀 점과 반영한 커밋을 함께 남겨주시고, 별도로 전체적인 피드백 반영 목록까지 표로 정리해주셔서 좋았습니다! 감사합니다. 🙇 
제가 드린 리뷰 내용들을 전체적으로 다 잘 이해하시고 적용해주신 것 확인했어요.
2단계에서 뵈어요! 💪
